Introduce an ArgsManager class encapsulating cs_args, mapArgs  and mapMultiArgs

jtimon opened the pull request:
Continues #9243 by @TheBlueMatt 
Introduce an ArgsManager class encapsulating cs_args, mapArgs  and mapMultiArgs.

To reduce disruption, the used functions are kept as wrapers. The wrapers can be removed later, see https://github.com/jtimon/bitcoin/compare/0.13-args-class...jtimon:0.13-args-remove-wrapers for the less disruptive ones to remove.

TheBlueMatt commented:
Concept ACK. I think we had discussed doing this when #9243 went in.

On January 9, 2017 11:02:04 AM EST, "Jorge Timón" <notifications@github.com> wrote:
>Continues #9243 by @TheBlueMatt 
>Introduce an ArgsManager class encapsulating cs_args, mapArgs  and
>mapMultiArgs.
>
>To reduce disruption, the used functions are kept as wrapers. The
>wrapers can be removed later, see
>https://github.com/jtimon/bitcoin/compare/0.13-args-class...jtimon:0.13-args-remove-wrapers
>for the less disruptive ones to remove.
>
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/bitcoin/bitcoin/pull/9494
>
>-- Commit Summary --
>
>  * Util: Create ArgsManager class
>  * Util: Expose ArgsManager argsGlobal
>  * Util: s/mapMultiArgs.count(/argsGlobal.IsArgSet(/
>  * Util: Introduce ArgsManager::ArgsAt()
>  * Util: Put mapMultiArgs inside ArgsManager
>
>-- File Changes --
>
>    M src/httprpc.cpp (4)
>    M src/httpserver.cpp (8)
>    M src/init.cpp (48)
>    M src/net.cpp (10)
>    M src/test/util_tests.cpp (77)
>    M src/util.cpp (83)
>    M src/util.h (28)
>
>-- Patch Links --
>
>https://github.com/bitcoin/bitcoin/pull/9494.patch
>https://github.com/bitcoin/bitcoin/pull/9494.diff
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/9494

jtimon commented:
Yes, we discussed it in #9243, but that this seems approximately what was discussed previously is still very useful, thanks.

What concerns me more of the current code is the fact that the methods IsArgSet, GetArg, GetBoolArg and  ArgsAt aren't const, because they need to ```LOCK(cs_args);```. The first "solution" that comes to mind is to make a superclass with just these methods (but they remain non-const).

dcousens commented:
concept ACK :+1:

jtimon commented on src/util.cpp:
```diff
@@ -377,41 +373,46 @@ void ParseParameters(int argc, const char* const argv[])
-bool IsArgSet(const std::string& strArg)
+const std::vector<std::string>& ArgsManager::ArgsAt(const std::string& strArg) const
+{
+    return mapMultiArgs.at(strArg);
```
I assume this should get LOCK(cs_args); too? If so, the method can't be const.

TheBlueMatt replied:
Yes, it also cant return a reference, needs to make a copy.

jtimon commented:
Needed rebase.

jtimon commented:
Needed rebase

laanwj commented:
Concept ACK, seems a move in the right direction.

sipa commented:
utACK c67209128a114e3260ab93c3e8a63b1b23663b34

TheBlueMatt commented:
Concept ACK, will review.

On March 24, 2017 2:57:16 PM PDT, Pieter Wuille <notifications@github.com> wrote:
>utACK c67209128a114e3260ab93c3e8a63b1b23663b34
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/9494#issuecomment-289152860

TheBlueMatt reviewed (commented):
Switching mapMultiArgs.count() to argsGlobal.IsArgSet makes a few things incorrect, because the SoftSetArgs stuff doesn't update mapMultiArgs but does update mapArgs. I think the correct solution here is to keep mapMultiArgs and mapArgs consistent (@luke-jr also needed that for the multiwallet pull), though that does mean ArgsAt needs to return a copy and not a reference (which I think is OK - arg usage doesn't need to be fast).

jtimon commented:
> Switching mapMultiArgs.count() to argsGlobal.IsArgSet makes a few things incorrect, because the SoftSetArgs stuff doesn't update mapMultiArgs but does update mapArgs.

Maybe a solution could be to use argsGlobal.IsArgsAt().size() > 0 in those cases?
I mean, I think in all those cases will just try to loop through an empty vector (ie do nothing) inside the if clause, but I haven't made sure of that. I could go through them and maybe replace some BOOST_FOREACH for C++ fors while at it.
Never mind, it seems mapMultiArgs.count("-bind") is not like that, but could be fixed with argsGlobal.IsArgsAt().size() > 0
By "making mapMultiArgs and mapArg" I assume you just mean redundantly setting to mapMultiArgs in ArgsManager::SoftSetArg too as it is done in ArgsManager::ArgsAt. Perhaps that is simpler and easier to reason about.
 
Thoughts?

TheBlueMatt commented:
I'd prefer to just fix the issue - I think its super strange that SoftSetArg, etc dont keep mapMultiArgs and mapArgs consistent.

jtimon commented:
Fixed nits. Also removed a bunch of BOOST_FOREACH for free in the commit that introduces ArgsManager::ArgsAt().

jtimon commented:
Fixed nit.

jtimon commented:
Rebased

JeremyRubin commented:
put my review of this code on the wrong PR. https://github.com/bitcoin/bitcoin/pull/10119#discussion_r109692600

jtimon commented:
Needed rebase again. Btw, should I squash the last commit in #10119 (everything else is just equal to this PR) to the first commit here or leave it separated?

jtimon commented:
Needed rebase again.

luke-jr commented on src/util.cpp:
```diff
@@ -414,63 +410,114 @@ void ParseParameters(int argc, const char* const argv[])
 }
 
-bool IsArgSet(const std::string& strArg)
+std::vector<std::string> ArgsManager::ArgsAt(const std::string& strArg)
```
`GetArgs` seems like a better name?

luke-jr commented on src/util.cpp:
```diff
@@ -414,63 +410,114 @@ void ParseParameters(int argc, const char* const argv[])
+    mapMultiArgs[strArg].push_back(strValue);
 }
 
+void ParseParameters(int argc, const char* const argv[])
```
Suggest moving these to the header as static inline.

jtimon replied:
I suspect a performance improvement is expected by doing so. Thanks

luke-jr commented on src/util.h:
```diff
@@ -226,6 +234,20 @@ bool SoftSetBoolArg(const std::string& strArg, bool fValue);
 void ForceSetArg(const std::string& strArg, const std::string& strValue);
+};
+
+extern ArgsManager argsGlobal;
```
Can we just call it `Args`? :)

jtimon replied:
I was thinking that maybe some functions could get a local ```ArgsManager& args``` parameters instead of using the global. Or do you mean s/ArgsManager/Args/ ? I'm fine with that.
Also, maybe s/argsGlobal/gArgs/ too, but I don't like ```extern ArgsManager args```

luke-jr replied:
I meant `extern ArgsManager Args`, but `gArgs` is probably good enough.

luke-jr commented on src/init.cpp:
```diff
@@ -734,7 +734,7 @@ void InitParameterInteraction()
     }
 
-    if (mapMultiArgs.count("-connect") && mapMultiArgs.at("-connect").size() > 0) {
+    if (argsGlobal.IsArgSet("-connect") && argsGlobal.ArgsAt("-connect").size() > 0) {
```
Under what conditions would `IsArgSet` be true, and `ArgsAt().size()` be zero?

jtimon replied:
After the last change suggested by @TheBlueMatt this should never be possible, good catch.

luke-jr commented on src/init.cpp:
```diff
@@ -904,9 +904,9 @@ bool AppInitParameterInteraction()
+    if (argsGlobal.IsArgSet("-debug")) {
         // Special-case: if -debug=0/-nodebug is set, turn off debugging messages
-        const std::vector<std::string>& categories = mapMultiArgs.at("-debug");
+        const std::vector<std::string>& categories = argsGlobal.ArgsAt("-debug");
```
Can you reference a temporary like this?

sipa replied:
You can. The lifetime of the temporary is extended by taking a reference to it. This code is effectively equivalent to:
```c++
const std::vector<std::string> categories_tmp = argsGlobal.ArgsAt("-debug");
const std::vector<std::string>& categories = categories_tmp;
```

jtimon replied:
At the same time I didn't do it on purpose here. I'm happy to remove the "&" if that's preferred.

luke-jr reviewed (commented):
Looks basically okay.

jtimon commented:
Fixed some nits. I made the wrappers static inline and removed the less disruptive ones (which include ParseParameters and ReadConfigFile) as I was doing in #10119 (just squashed it to the first commit here, plus adapted 1 new use of SoftSetArg after rebase), but since that's a little bit more disruptive I didn't made the methods ParseParameters and ReadConfigFile inline yet. Also found more cases where ArgsAt/GetArgs didn't need to be called after calling IsArgSet.

jtimon committed:
Util: Create ArgsManager class...

- Introduce ArgsManager::GetArgs()
- Adapt util_tests.cpp to ArgsManager

jtimon committed:
Util: Put mapMultiArgs inside ArgsManager

- Set ArgsManager::mapMultiArgs in ArgsManager::SoftSetArg, ForceSetArg, SoftSetBoolArg

jtimon committed:
scripted-diff: Util: Encapsulate mapMultiArgs behind gArgs

-BEGIN VERIFY SCRIPT-
sed -i 's/mapMultiArgs.count(/gArgs.IsArgSet(/g' ./src/*.h ./src/*.cpp ./src/*/*.h ./src/*/*.cpp ./src/*/*/*.h ./src/*/*/*.cpp ;
sed -i 's/mapMultiArgs.at("/gArgs.GetArgs("/g' ./src/*.h ./src/*.cpp ./src/*/*.h ./src/*/*.cpp ./src/*/*/*.h ./src/*/*/*.cpp ;
-END VERIFY SCRIPT-

jtimon committed:
Util: Small improvements in gArgs usage

- Don't check gArgs.IsArgSet() is greater than 0
- Remove unneeded calls and local variables

jtimon commented:
Renamed s/globalArgs/gArgs/ as suggested discussed with @luke-jr here. @TheBlueMatt asked for using scripted-diff so I ended up "bikeshedding the history" quite a bit, but I think it's clearer now.
I tried to follow @luke-jr 's suggestion https://github.com/bitcoin/bitcoin/pull/9494#discussion_r114871261 in more places, but I don't understand why this breaks the python tests yet: https://github.com/jtimon/bitcoin/commit/32bffa0162a0ebd66c554dab4b35705e664845dc

Also, sorry for going back and forth with removing some wrappers directly or not, but with scripted-diff I think it's better to just remove them all at once after this. But if we want to do the simpler ones without scripted-diff here I still have that in https://github.com/jtimon/bitcoin/tree/b15-args-class-wrappers

TheBlueMatt commented on src/util.cpp:
```diff
@@ -420,57 +418,64 @@ void ParseParameters(int argc, const char* const argv[])
 {
     LOCK(cs_args);
     mapArgs[strArg] = strValue;
+    _mapMultiArgs[strArg].push_back(strValue);
```
Its somewhat annoying to have an intermediate commit which breaks our locking in the middle - maybe move this one line to the "Put mapMultiArgs inside ArgsManager" commit? It should at least break fewer things by doing so.

TheBlueMatt reviewed (commented):
utACK f19f5b5dbd0e3b6f73ded3818014eed3f728e5fc

ryanofsky commented on src/util.h:
```diff
@@ -235,6 +243,55 @@ bool SoftSetBoolArg(const std::string& strArg, bool fValue);
 
 // Forces a arg setting, used only in testing
 void ForceSetArg(const std::string& strArg, const std::string& strValue);
+};
```
"Introduce an ArgsManager class encapsulating cs_args, mapArgs and mapMultiArgs"

Indentation for all the methods above is missing. Though it's understandable not to add it now to make diff smaller.

jtimon replied:
Yes, that's exactly the point, but I have done similar things in the past and not fixed them later, something I feel bad about, for example: https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L1313
Please, help me make sure that it gets done this time.

ryanofsky reviewed (commented):
utACK f19f5b5dbd0e3b6f73ded3818014eed3f728e5fc

theuni commented on src/init.cpp:
```diff
@@ -911,9 +911,9 @@ bool AppInitParameterInteraction()
+    if (gArgs.IsArgSet("-debug")) {
         // Special-case: if -debug=0/-nodebug is set, turn off debugging messages
-        const std::vector<std::string>& categories = gArgs.GetArgs("-debug");
+        const std::vector<std::string> categories = gArgs.GetArgs("-debug");
```
why?

jtimon replied:
Because https://github.com/bitcoin/bitcoin/pull/9494#discussion_r114871456 and I don't care either way for the reasons @sipa explained.

theuni commented:
utACK f19f5b5dbd0e3b6f73ded3818014eed3f728e5fc

jtimon commented:
Sorry, just pushed again changing the history as @TheBlueMatt suggested. 
78da882 should be equivalent to f19f5b5

ryanofsky commented:
utACK 78da882edd54ed62f4a0035590460a97cb9ff282. Confirmed no diff since previous review, just mapMultiArgs commit history change.

laanwj commented:
utACK 78da882
